### ISSUE TITLE
[RFC] HiCache stage 2: HiCacheController — L1↔L2 async pipeline

### ISSUE THREAD

**Issue by cjx0709** [sgl-project/sglang-jax#1312]
## Summary

Implement `HiCacheController`, the control plane for L1 (HBM) <-> L2 (host pinned) KV page transfers:

- **D2H write** (async): when a tree cache node is reused (hit), asynchronously copy its KV from L1 to L2
- **H2D load** (sync): when prefill discovers an L1 miss but L2 hit, synchronously copy from host pool back to device pool
- **Eviction coordination**: when L2 is full, the tree cache layer initiates eviction, and the controller executes slot release (this stage does not connect L3 — eviction discards data)

> **Interface alignment** — Transfers go exclusively through the `HostKVPool` `buffer_id`-addressed primitives (`copy_into` / `copy_to_device`); the controller does **not** call `jax.device_put` itself. The A/B transfer format is now an implementation detail of the pool, not the controller. See the decision record `docs/design/hicache_pd_raiden_alignment.md` (D4/D5). The shared ABC (#1368) is `reserve()/release(buffer_id)/copy_from_device(layers, buffer_id)`; HiCache reserves a slot and fills it via `copy_into`. This keeps the controller raiden-switchable with zero changes (see §8).

## Position in the roadmap

1. Host pool foundation
2. UnifiedRadixCache + FullComponent port
3. **HiCacheController + async pipeline** *(this issue)*
4. End-to-end integration and validation
5. L3 persistent storage
6. Multi-component + tuning

Depends on Stage 0 (`LRUHostKVPool` + ABC transfer primitives) and Stage 1 (UnifiedRadixCache + FullComponent).

## Proposed structure changes

- `sgl_jax/srt/mem_cache/hicache_controller.py` — **new**: `HiCacheController`
- `sgl_jax/test/mem_cache/test_hicache_controller.py` — **new**: unit tests
- `sgl_jax/test/disagg_feasibility/t_hicache_d2h_format.py` — **new**: D2H format benchmark (spike; decides the `LRUHostKVPool` internal format, not the controller interface)
- `sgl_jax/test/disagg_feasibility/t_hicache_host_lifecycle.py` — **new**: host memory lifecycle validation (spike)

### 1. Data organization

**Device side (L1)**

The device pool (`MHATokenToKVPool`) stores per-layer. `kv_buffer` is `list[jax.Array]` of length `num_layers`, each with shape:

```text
kv_buffer[layer_id]: (num_pages, page_size, kv_heads*2//packing, packing, head_dim)
                      ^^^^^^^^                ^^^^^^^^^^^^^^^^^^^
                      DP shard ("data")       TP shard ("tensor")
```

**Host side (L2)**

`LRUHostKVPool` (Stage 0) manages host pinned slots. The controller **never touches** a slot's internal `jax.Array` — it holds only `int buffer_id` and moves data via `host_pool.copy_into` / `copy_to_device`. Host layout (per-slot single array vs list, all-layers-packed or not) is a pool implementation detail invisible to the controller.

**D2H / H2D transfer format (pool-internal detail)**

The transfer format (all-layers-packed vs per-layer direct) is **fully localized inside `LRUHostKVPool.copy_into` / `copy_to_device` (D5)** and does not enter the controller interface:

- **A (all-layers-packed):** pool-internal `jnp.stack` then bulk move; fewer calls, temp HBM overhead.
- **B (per-layer direct):** pool-internal per-layer move; no temp HBM, more calls.

Final choice is decided by the Stage 0 `LRUHostKVPool` benchmark. **Regardless of choice, the controller interface is unchanged**: it always calls `copy_into([device_idx], [buffer_id])` / `copy_to_device([buffer_id], [device_idx])`. Under raiden the host layout is engine-decided, making this moot for the controller.

### 2. Interface

```python
class HiCacheController:
    def __init__(self, host_pool: HostKVPool, device_pool: KVCache,
                 device_allocator: BaseTokenToKVPoolAllocator):
        self._executor = ThreadPoolExecutor(max_workers=1)
        self._pending: list[Future] = []

    def write(self, device_indices, host_buffer_ids) -> None:
        """Async D2H: copy device page(s) into host pool slot(s).
        device_indices[i] pairs with host_buffer_ids[i] (int buffer_id from reserve())."""

    def load(self, host_buffer_ids, device_indices) -> None:
        """Sync H2D: copy host pool slot(s) back to device pool page(s)."""

    def evict_callback(self, host_buffer_ids) -> None:
        """L2 eviction callback. This stage frees slots; Stage 4 connects L3."""

    def check_write_status(self) -> None:
        """Non-blocking poll: clean up completed D2H futures, release node locks."""

    def drain_pending(self) -> None:
        """Blocking wait for all pending D2H (e.g. before L1 eviction, shutdown)."""

    def pending_count(self) -> int: ...
```

`write`/`load` **do not call `jax.device_put` themselves** — they delegate to the `HostKVPool` `buffer_id`-addressed primitives:

```python
def _do_d2h(self, device_indices, host_buffer_ids):
    self._host_pool.copy_into(list(device_indices), list(host_buffer_ids))

def _do_h2d(self, host_buffer_ids, device_indices):
    self._host_pool.copy_to_device(list(host_buffer_ids), list(device_indices))
```

This is where the switchability contract (D4) lands in the controller: **the controller only ever sees `int buffer_id`, never a host `jax.Array`.** The actual movement (`jax.device_put`, host layout, format A/B) is sealed inside the pool. It does not reuse `KVCache.get_cpu_copy` / `load_cpu_copy` (which go `jax.device_get` -> numpy, non-pinned, non-async).

### 3. Async execution model

D2H uses `ThreadPoolExecutor` (1 worker). JAX's `device_put` releases the GIL at the C++ level, so D2H can run in parallel with the main thread's forward.

```python
def write(self, device_indices, host_buffer_ids):
    future = self._executor.submit(self._do_d2h, device_indices, host_buffer_ids)
    self._pending.append(future)

def check_write_status(self):
    done = [f for f in self._pending if f.done()]
    for f in done:
        f.result()  # raise exceptions if any
        self._pending.remove(f)

def drain_pending(self):
    for f in self._pending:
        f.result()
    self._pending.clear()
```

H2D (`load`) runs synchronously on the main thread — ~45.7 GB/s, microsecond-range per page. It also delegates to `host_pool.copy_to_device`; the controller does no direct `device_put`.

### 4. D2H write strategy

D2H is not triggered on every insert. `UnifiedRadixCache` maintains `hit_count` per node; when a node is **reused** (`match_prefix` hit) and `hit_count >= write_through_threshold`, `write_backup` (D2H) triggers.

```text
insert creates new node -> hit_count = 0, no D2H
subsequent match_prefix hit -> hit_count += 1
hit_count >= threshold -> _inc_hit_count triggers write_backup -> controller.write()
```

`write_through_threshold` defaults to 1. This means: decode does not D2H every step; only reused pages are worth D2H; partially-filled pages do not D2H.

After submission nodes are locked; `check_write_status()` polls completion, then releases locks and marks `node.backuped = True` (called before `get_new_batch_prefill`).

### 5. H2D load timing

H2D (prefetch) occurs during prefill batch construction. When `match_prefix` returns `last_host_node` (L1 miss but L2 hit):

```text
get_new_batch_prefill
  +- check_write_status()  -> non-blocking poll D2H completion
  +- req.init_next_round_input -> match_prefix
       +- if host_hit_length > 0:
            +- device_indices = allocator.alloc(host_hit_length)
            +- controller.load(host_buffer_ids, device_indices)  -> sync H2D
            +- update tree node: value = device_indices, host_value = None
```

### 6. Step-level timing overview

```text
Step N-1 forward completes
  +- process_batch_result -> insert -> _inc_hit_count -> threshold
       +- write_backup -> controller.write(device_idx, [buffer_id]) -> async D2H
Step N: get_next_batch_to_run
  +- check_write_status()  -> non-blocking poll
  +- get_new_batch_prefill -> match_prefix -> [if L2 hit] controller.load() -> sync H2D
  +- run_batch -> forward (D2H runs async in background)
```

**D2H latency estimates** (Qwen3-8B, page_size=1, bf16):

| Scenario | Pages | Size | D2H (@2.8 GB/s) | Forward |
|---|---|---|---|---|
| decode step (single page hit) | 1 | ~128 KB | ~0.045 ms | ~10-20 ms |
| prefill 64 tokens hit | 64 | ~8 MB | ~2.9 ms | ~15-30 ms |
| prefill 2048 tokens hit | 2048 | ~256 MB | ~91 ms | ~200-500 ms |

`drain_pending` is only called at consistency-critical points.

### 7. Lifecycle

```text
init_hicache():
  host_pool = LRUHostKVPool(device_pool, pool_size, page_size, layer_num, per_layer_shape, dtype, mesh, partition_spec)
  controller = HiCacheController(host_pool, device_pool, allocator)

Running:
  insert -> _inc_hit_count -> threshold:
    buffer_id = host_pool.reserve()                -> buffer_id
    controller.write(device_indices, [buffer_id])  -> async D2H
    tree_node: lock, host_value = buffer_id
  check_write_status(): D2H done -> unlock, node.backuped = True
  match_prefix -> L1 miss, L2 hit:
    controller.load(host_buffer_ids, device_indices)  -> sync H2D
  L2 full (reserve returns None):
    FullComponent.drive_host_eviction -> controller.evict_callback([buffer_id])
      -> host_pool.release(buffer_id); tree_node.host_value = None
Shutdown: controller.drain_pending()
```

Throughout, the upper layer sees only `int buffer_id`; no host `jax.Array` is exposed.

### 8. Switchability: raiden migration

The controller's contract (D4): **only `int buffer_id` + device index, moving via `host_pool.copy_into` / `copy_to_device`, never touching a host `jax.Array` or calling `jax.device_put` directly.** So at raiden time:

| Change | Location | Controller impact |
|---|---|---|
| Add `RaidenHostKVPool(HostKVPool)`, `copy_into→raiden.d2h`, `copy_to_device→raiden.h2d` | Stage 0 (host pool backend) | **zero** |
| `init_hicache` swaps `LRUHostKVPool` → `RaidenHostKVPool` | wiring/config | **zero** |

`HiCacheController`, `FullComponent` hooks, `UnifiedRadixCache` are all untouched.

### Spike 1: D2H transfer format benchmark

Compare approach A (all-layers-packed) vs B (per-layer direct) on v6e-4 (Qwen3-8B: 32 layers, 8 kv_heads, 128 head_dim, bf16). **This decides the `LRUHostKVPool` internal format (Stage 0), not the controller interface** — the controller always uses the `buffer_id` primitives.

1. Single page D2H latency (median of 100); 2. Batch D2H (64/256 pages); 3. Device HBM peak increment; 4. D2H + forward overlap.

| Metric | Choose A if | Choose B if |
|---|---|---|
| Single page D2H latency | A >30% lower | A no advantage / slower |
| HBM peak increment | stack temp < 1% total HBM | unacceptable |
| Overlap | both overlap | — |

### Spike 2: Host pinned memory lifecycle validation

Verify managing host pinned memory via JAX (create via `device_put`, reclaim via GC) works: reference hold/release, batch release, repeated alloc/free (1000x, RSS bounded), multi-slot coexistence, long-term data integrity. Key concerns: delayed reclamation/fragmentation for `pinned_host`, TPU runtime alloc/free frequency limits, many small vs few large array GC behavior. Env: v6e-4. Code in `sgl_jax/test/disagg_feasibility/`.

## Acceptance criteria

- **Spike: D2H format** — A/B benchmark complete with clear data
- **Spike: host memory lifecycle** — all items pass
- write correctness: D2H then `copy_to_device` round-trip data allclose with device original
- load correctness: H2D then device KV buffer allclose with host original
- Does not expose jax.Array: controller public method signatures contain only `int buffer_id` / device index
- Async: `write` returns immediately without blocking
- check_write_status / drain_pending: correct poll / drain behavior
- Overlap: forward + D2H total time < forward + D2H
- Round-trip: D2H -> H2D data allclose with original
- evict_callback: after eviction, slot freed, `available_size()` increases
